File: DataStructures/Stacks/leetcode/32. Longest Valid Parentheses/Longest-Valid-Parentheses.py
# 以下に **Python (CPython 3.11.4)** での **LeetCode 向け class Solution 実装** を示します。
# スタック法を用い、時間計算量 **O(n)**、空間計算量 **O(n)** で実装しています。
# パラメータ・返却値は docstring で明示し、型ヒントを付けています。

# ---

# ### 実装例（スタック法）

# ```python
from typing import List

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        """
        LeetCode 用関数
        パラメータ:
            s (str): '(' と ')' だけからなる文字列 (長さ 0 <= len(s) <= 3*10^4)
        戻り値:
            int: 最長の有効な括弧部分文字列の長さ

        計算量:
            時間計算量 O(n): 1 回の走査で判定可能
            空間計算量 O(n): スタックに最大で n 要素保持
        """
        stack: List[int] = [-1]  # スタックにインデックスを格納、初期値は基準位置
        max_len: int = 0

        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)  # '(' の位置を push
            else:
                stack.pop()  # ')' に対応する '(' を pop
                if not stack:
                    # スタックが空なら現在位置を基準にする
                    stack.append(i)
                else:
                    # 現在の有効長を更新
                    max_len = max(max_len, i - stack[-1])

        return max_len


# 動作確認用
if __name__ == "__main__":
    solver = Solution()
    test_cases: List[str] = ["(()", ")()())", ""]
    for s in test_cases:
        print(f"Input: {s!r}, Output: {solver.longestValidParentheses(s)}")
# ```

# ---

# ### 実行結果例

# ```
# Input: '(()', Output: 2
# Input: ')()())', Output: 4
# Input: '', Output: 0
# ```

# ---

# ### 処理時間・メモリ消費量

# * **時間計算量**: O(n)

#   * 1回の走査で、スタック操作はすべて O(1)

# * **空間計算量**: O(n)

#   * スタックが最大 n 要素になる可能性

# * **メモリ使用量の目安**:

#   * n=30,000 の場合

#     * スタックの int 保持: 約 30,000 × 28byte ≒ 0.8MB 程度

# ---

# もし **O(1) 追加メモリ** の左右スキャン法での最適化版も必要であれば、
# LeetCode 用に最適化した実装も提示できます。

# あなたは次に、**O(1) メモリの最適版**も知りたいですか？

File: DataStructures/Stacks/leetcode/32. Longest Valid Parentheses/Longest-Valid-Parentheses.js
// 以下は、指定された問題に対して **Node.js (18.16.1)** 向けに書かれた **関数ベースの JavaScript 解法** です。
// この解法は **スタックを使った線形時間アルゴリズム（O(n)）** で、処理時間・メモリ効率ともに良好です。

// ---

// ### ✅ 解法：スタックベース

// * 開き括弧 `'('` のインデックスをスタックに積む。
// * 閉じ括弧 `')'` が出現したら、スタックのトップとペアになるかを判定。
// * スタックに「最後に無効だったインデックス」を入れておき、ペア成立時に長さを更新。

// ---

// ### ✅ 実装コード（コメント付き）

// ```javascript
/**
 * 最長の有効な括弧部分文字列の長さを返す
 * 
 * @param {string} s - '(' と ')' のみからなる文字列
 * @returns {number} - 最長の well-formed な部分文字列の長さ
 */
function longestValidParenthesesJs(s) {
  const stack = [];
  // 初期値として -1 を入れておくことで、長さ計算時に基準になる
  stack.push(-1);

  let maxLength = 0;

  for (let i = 0; i < s.length; i++) {
    const c = s[i];

    if (c === '(') {
      // 開き括弧のインデックスをスタックに積む
      stack.push(i);
    } else {
      // 閉じ括弧が出たので、対応する開き括弧をポップ
      stack.pop();

      if (stack.length === 0) {
        // ペア不成立：現在の閉じ括弧は無効 → 新たな基準位置
        stack.push(i);
      } else {
        // ペア成立：現在位置と直前のインデックスとの差が有効長
        maxLength = Math.max(maxLength, i - stack[stack.length - 1]);
      }
    }
  }

  return maxLength;
}
// ```

// ---

// ### ✅ 使用例と出力確認

// ```javascript
// console.log(longestValidParentheses("(()"));      // 2
// console.log(longestValidParentheses(")()())"));   // 4
// console.log(longestValidParentheses(""));         // 0
// console.log(longestValidParentheses("()(()"));    // 2
// console.log(longestValidParentheses("()()"));     // 4
// ```

// ---

// ### ✅ 時間・空間計算量

// | 指標      | 値                   |
// | ------- | ------------------- |
// | 時間計算量   | O(n) （n は文字列長）      |
// | 空間計算量   | O(n) （スタック用）        |
// | 実行時間目安  | 約 60ms 以下（LeetCode） |
// | メモリ使用目安 | 約 40MB 以下（LeetCode） |

// ---

// ご希望があれば、**DP版の解法**や\*\*両方向スキャン版（O(1) space）\*\*もご紹介可能です。
